coderd/keyrotate/rotate_internal_test.go

```diff
@@ -0,0 +1,526 @@
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features: []database.CryptoKeyFeature{
+				database.CryptoKeyFeatureWorkspaceApps,
+			},
+		}
+
+		now := dbnow(clock)
+
+		// Seed the database with an existing key
+		deletingKey := dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature:  database.CryptoKeyFeatureWorkspaceApps,
+			StartsAt: now,
+			Sequence: 456,
+			DeletesAt: sql.NullTime{
+				Time:  now.Add(time.Hour),
+				Valid: true,
+			},
+		})
+
+		// We should only have inserted a key.
+		err := kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		keys, err := db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 2)
+		oldKey, newKey := keys[0], keys[1]
+		if oldKey.Sequence != deletingKey.Sequence {
+			oldKey, newKey = newKey, oldKey
+		}
+		requireKey(t, oldKey, deletingKey.Feature, deletingKey.StartsAt.UTC(), deletingKey.DeletesAt.Time.UTC(), deletingKey.Sequence)
+		requireKey(t, newKey, deletingKey.Feature, now, time.Time{}, deletingKey.Sequence+1)
+	})
+
+	t.Run("NoKeys", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features: []database.CryptoKeyFeature{
+				database.CryptoKeyFeatureWorkspaceApps,
+			},
+		}
+
+		err := kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		keys, err := db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 1)
+		requireKey(t, keys[0], database.CryptoKeyFeatureWorkspaceApps, clock.Now().UTC(), time.Time{}, 1)
+	})
+
+	// Assert we insert a new key when the only key was manually deleted.
+	t.Run("OnlyDeletedKeys", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features: []database.CryptoKeyFeature{
+				database.CryptoKeyFeatureWorkspaceApps,
+			},
+		}
+
+		now := dbnow(clock)
+
+		deletedkey := dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature:  database.CryptoKeyFeatureWorkspaceApps,
+			StartsAt: now,
+			Sequence: 19,
+			DeletesAt: sql.NullTime{
+				Time:  now.Add(time.Hour),
+				Valid: true,
+			},
+			Secret: sql.NullString{
+				String: "deleted",
```

Comment:
strange that we're setting the string in an invalid string --- I'm guessing this is to fool `dbgen` but it would be good to comment